### ISSUE TITLE
Laravel 5.3 $request->route() is null in Middleware 

### ISSUE THREAD

**Issue by Mr-Anonymous** [laravel/framework#16098]
- Laravel Version: 5.3.19
- PHP Version: 5.6.10
- Database Driver & Version: Mysql

### Description:
I have created a Middleware that needs to access the api route's url parameter. But since upgrading Laravel from 5.1 to 5.3, the `request->route('parameter_name')` is not working and always returns as null.

### Steps To Reproduce:
This is my set-up:

**routes/api.php:**
```
Route::group(['middleware' => ['app.access']], function()
{
    Route::resource('apps/{apps}/houses', 'HouseController',
                    ['except' => ['index', 'create', 'edit']]);

    Route::get('apps/{apps}/houses/colour/{colour}', 'HouseController@colourCheck');
    ...
}
```

**Kernel.php has `$routeMiddleware` updated like this:**
```
protected $routeMiddleware = [
        'auth' => \Illuminate\Auth\Middleware\Authenticate::class,
        'auth.basic' => \Illuminate\Auth\Middleware\AuthenticateWithBasicAuth::class,
        'bindings' => \Illuminate\Routing\Middleware\SubstituteBindings::class,
        'can' => \Illuminate\Auth\Middleware\Authorize::class,
        'guest' => \App\Http\Middleware\RedirectIfAuthenticated::class,
        'throttle' => \Illuminate\Routing\Middleware\ThrottleRequests::class,
        // Added ones....
        'jwt.auth' => \Tymon\JWTAuth\Middleware\GetUserFromToken::class,
        'jwt.refresh' => \Tymon\JWTAuth\Middleware\RefreshToken::class,
        // Custom Middleware
        'app.access' => \App\Http\Middleware\AppAccess::class,
    ];
```

**Middleware:**
```
public function handle($request, Closure $next)
{
        $appId = $request->route('apps');
        ...
}
```

I am getting the following response from Middleware:

1) `$request->route()` is `null`

2) `$request->route('apps)` is `null`

3) `$request->route()->parameter('apps')` is `null`

3) Below gives `Undefined index: apps`:
```
$parameters = $request->route()->parameters();
$appId = $parameters['apps'];
```

4) But when I `print_r($parameters);` in the above code, the output is like this:
```
Array
(
    [apps] => 1
    [day] => tuesday
)
[]
```

I did not have this issue with Laravel 5.1 when the middleware had the same set-up. But since updating the Laravel 5.3 and changing the routes to api and web, I am unable to get the appId from the Middleware. Has something changed or am I missing something?

**Comment by themsaid** [comment]
`$request->route()` is null while `$request->route()->parameters()` has values? How come?

**Comment by Mr-Anonymous** [comment]
Thats exactly right and thats what I don't understand. Similarly, `$request->route()->parameters()` has values but `$request->route()->parameter('apps')` is null.

**Comment by Mr-Anonymous** [comment]
And another weird thing I noticed is, when I `print_r($request->route()->parameters());` from the Middleware, I get different parameter name for the same routes. For example: 
 
Route `Route::resource('apps/{apps}/houses', 'HouseController')` returns:
```
Array
(
    [app] => 1
)
```
Route `Route::get('apps/{apps}/houses/colour/{colour}', 'HouseController@colourCheck')` returns:
```
Array
(
    [apps] => 1
    [day] => tuesday
)
```
I don't understand how the first one had the route parameter name as 'app' instead of 'apps'.

**Comment by Mr-Anonymous** [comment]
Since my last post, I now got it working by changing all `{apps}` url parameter in my api route file to `{app}`. After that, I updated the Middleware to get the parameter using `$request->route()->parameter('app')`. It seems to work after this change. However, I still find it odd that `$request->route()` was returning `null` whereas `$request->route()->parameters()` had valued and why the `resource` in route was using `app` as parameter name instead of `apps` which was mentioned in the route file.

**Comment by themsaid** [comment]
PHP doesn't work like this, I actually couldn't replicate the issue in a fresh 5.3 installation using your steps to reproduce, calling a function on null would return error.

**Comment by Mr-Anonymous** [comment]
You are right, the reason for the mix up was because the resource routes were generating route url parameter as `app` instead of `apps`. As a result some routes were returning `null` for the parameter name `apps` and some where showing the parameter. I combined both outputs and misunderstood the behaviour. Once I renamed all route parameters to `{app}` in the `routes/api.php` and Middleware, the Middleware started working the way it should.

Having said that, I don't know why the resource routes were creating the parameter as `app` when I have added the parameter as `{apps}` in route file, but the above edit fixed the error for me, atleast for now.
